contributor-docs/code-change-guide.md

```diff
@@ -0,0 +1,518 @@
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+This guide is for Beam users and developers changing and testing Beam code.
+Specifically, this guide provides information about:
+
+1. Testing code changes locally
+
+2. Building Beam artifacts with modified Beam code and using it for pipelines
+
+# Repository structure
+
+The Apache Beam GitHub repository (Beam repo) is pretty much a "mono repo",
+containing everything in the Beam project, including the SDK, test
+infrastructure, dashboards, the [Beam website](https://beam.apache.org),
+the [Beam Playground](https://play.beam.apache.org), and so on.
+
+## Gradle quick start
+
+The Beam repo is a single Gradle project that contains all components, including Python,
+Go, the website, etc. It is useful to familiarize yourself with the concept of Gradle project structure:
+https://docs.gradle.org/current/userguide/multi_project_builds.html
+
+### Gradle key concepts
+
+* project: folder with build.gradle
+* task: defined in build.gradle
+* plugin: run in project build.gradle, pre-defined tasks and hierarchies
+
+For example, common tasks for a Java project or subproject include:
+
+- compileJava
+- compileTestJava
+- test
+- integrationTest
+
+To run a Gradle task, the command is `./gradlew -p <project path> <task>` or equivalently, `./gradlew :project:path:task_name`. For example:
+
+```
+./gradlew -p sdks/java/core compileJava
+
+./gradlew :sdks:java:harness:test
+```
+
+### Gradle project configuration: Beam specific
+
+* A **huge** plugin `buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin` manages everything.
+
+Then, in each java project or subproject, the `build.gradle` file starts with:
+
+```groovy
+
+apply plugin: 'org.apache.beam.module'
+
+applyJavaNature( ... )
+```
+
+Relevant usage of `BeamModulePlugin` includes:
+* Manage Java dependencies
+* Configuring projects (Java, Python, Go, Proto, Docker, Grpc, Avro, an so on)
+  * Java -> applyJavaNature; Python -> applyPythonNature, and so on
+  * Define common custom tasks for each type of project
+    * `test`: run Java unit tests
+    * `spotlessApply`: format java code
+
+## Code paths
+
+The following are example code paths relevant for SDK development:
+
+* `sdks/java` Java SDK
+  * `sdks/java/core` Java core
+  * `sdks/java/harness` SDK harness (entrypoint of SDK container)
+
+* `runners` runner supports, written in Java. For example,
+  * `runners/direct-java` Java direct runner
+  * `runners/flink-java` Java Flink runner
+  * `runners/google-cloud-dataflow-java` Dataflow runner (job submission, translation, etc)
+    * `runners/google-cloud-dataflow-java/`worker Worker on Dataflow legacy runner
```

Comment:
I think there's a typo in this line (worker Worker), but I'm not sure what it's supposed to be.

Reply:
It should read `runners/google-cloud-dataflow-java/worker`, will fix